secure_storage: add a ZMS-based implementation of the ITS store module

tomi-font opened the pull request:
See the commit messages.

A follow-up PR will introduce the (static) allocation of `psa_storage_uid_t` values for end users of the PSA Secure Storage API.

tomi-font committed:
secure_storage: add a ZMS-based implementation of the ITS store module

It becomes the new default when the secure_storage_its_partition
devicetree chosen property is defined as it is a preferred alternative.

See the help message of the
`CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_ZMS` Kconfig option
for more information.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

Laczen commented:
@tomi-font,
a. do not use `storage_partition`, this is already used for other purposes and cannot be used by two systems at the same time,
b. zms, nvs, or settings are not really a good fit for secure storage as they use more flash pages (at least 2) and ensure persistence, something that might not be required for secure storage,
c. the construction of the zms uid is fragile and might break as a result of firmware updates,

tomi-font committed:
tests: secure_storage: psa: its: add a test for the ZMS ITS store module

Add a test scenario for the ITS store module implementation that uses
ZMS for storage.
Reorganize the others at the same time.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
secure_storage: its: store: settings: stop using SETTINGS_MAX_VAL_LEN

Remove the hard restriction on CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE.
SETTINGS_MAX_VAL_LEN is in practice not used by any settings backend
implementation.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font commented:
> a. do not use storage_partition, this is already used for other purposes and cannot be used by two systems at the same time

Good point. I was thinking about how to handle potential conflicts, but I think that I will just remove the possibility to use `storage_partition`.

> b. zms, nvs, or settings are not really a good fit for secure storage as they use more flash pages (at least 2) and ensure persistence, something that might not be required for secure storage

This makes no sense.

> c. the construction of the zms uid is fragile and might break as a result of firmware updates

No it's not. And related to that, I will address the topic of UID assignment (for secure storage) in a follow-up PR.

> I would advise on the following alternative approach:
> [...]

I don't want to reinvent the storage wheel, plus I don't see the advantage in what you're proposing. Might as well just use the settings-based implementation.
For what you're proposing, there would need to be a mechanism to allocate ID ranges to different users of a same partition, which we don't currently have.

Laczen commented:
> > b. zms, nvs, or settings are not really a good fit for secure storage as they use more flash pages (at least 2) and ensure persistence, something that might not be required for secure storage
> 
> This makes no sense.
>
Ok if you think they are a good fit. 
> > c. the construction of the zms uid is fragile and might break as a result of firmware updates
> 
> No it's not. And related to that, I will address the topic of UID assignment (for secure storage) in a follow-up PR.
>
No need to address this in a follow-up PR, do it now before it gets used.
 
> > I would advise on the following alternative approach:
> > [...]
> 
> I don't want to reinvent the storage wheel, plus I don't see the advantage in what you're proposing. Might as well just use the settings-based implementation. For what you're proposing, there would need to be a mechanism to allocate ID ranges to different users of a same partition, which we don't currently have.

There are already users that are using the NVS backend for both settings and data storage, the allocation of ID ranges could be formalized.

If on the other hand it is needed that the secure storage has its own backend (as implied by the PR), a slightly modified copy of the flash backend used in tf-m would be more appropriate.

If the settings-based implementation fits the need, why propose a different alternative?

tomi-font commented:
> No need to address this in a follow-up PR, do it now before it gets used.

What I am talking about is not specific to what is added in this PR. It's about the PSA Secure Storage API overall. And the rationale for not including it as part of this PR is to keep the scope of a PR focused on one topic.

> There are already users that are using the NVS backend for both settings and data storage, the allocation of ID ranges could be formalized.

I agree that it could and even should be formalized.

> If on the other hand it is needed that the secure storage has its own backend (as implied by the PR)

No, it's not needed. This PR is just to allow direct use of a partition for more control and efficiency of the storage. For the majority of cases it will actually still default to using settings because `storage_partition` will be defined but not `secure_storage_partition`.

Laczen commented:
> No, it's not needed. This PR is just to allow direct use of a partition for more control and efficiency of the storage. For the majority of cases it will actually still default to using settings because `storage_partition` will be defined but not `secure_storage_partition`.

If it is worth using a partition for more control and efficiency of storage it would be good to target as much users as possible. NVS is still much broader used (it is more efficient for most users) and has a large install base. It doesn't seem a good choice to leave these users without a possible more efficient storage solution.

Splitting up the storage into metadata (uid, caller uid, flags, (optional) nonce, (optional) tag) and the data itself would give the following advantages:
1. No limitation is build into the uid,
2. No limitation is build into the caller uid,
3. The stack usage would be reduced as it is no longer required to create the combination of metadata and data before writing to storage,
4. It can be used by both nvs and zms,

tomi-font committed:
fs: zms: fix format string

Use %zu for size_t.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

rghaddab reviewed (commented):
Thank you for adding ZMS backend to secure_storage.
Some few comments but all minors.
Also maybe add a separate commit where you clang-format old code to get rid of github warnings

tomi-font commented on subsys/secure_storage/src/its/store/zms.c:
```diff
@@ -0,0 +1,128 @@
+SYS_INIT(init_zms, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+/* Bit position of the ITS caller ID in the ZMS entry ID. */
+#define ITS_CALLER_ID_POS 30
```
No, it should very much be a fixed value. This is an implementation detail (trick) that there are caller IDs introduced to the PSA ITS API. They won't grow in number, and they shouldn't be visible to the end users.

tomi-font commented on subsys/secure_storage/src/its/store/zms.c:
```diff
@@ -0,0 +1,128 @@
+
+static uint32_t zms_id_from(secure_storage_its_uid_t uid)
+{
+	return (uint32_t)uid.uid | (uid.caller_id << ITS_CALLER_ID_POS);
```
I had a look at `FIELD_PREP()` but to me its usage (having to pass a mask) would make it less clear than how it's like currently.

tomi-font commented on subsys/secure_storage/src/its/store/zms.c:
```diff
@@ -0,0 +1,128 @@
+
+static bool has_forbidden_bits_set(secure_storage_its_uid_t uid)
+{
+	if (uid.uid & ~(psa_storage_uid_t)((1 << ITS_CALLER_ID_POS) - 1)) {
```
True. I was just not really aware of those macros.

tomi-font commented:
>  Also maybe add a separate commit where you clang-format old code to get rid of github warnings

I'm not a fan of the suggestions. Plus they're just notices. (And have sparked some debate on Discord.) They can be hidden with the `a` key. :slightly_smiling_face:

tomi-font committed:
secure_storage: warn when there is no ITS store module implementation

Output a CMake error when the ITS store module is enabled but no
implementation ended up enabled (due to unfulfilled prerequisites).
This is to make it more clear than undefined references at link time.
Not a fatal error because CMake cannot fail for the twister filtering
to work on the tests.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
secure_storage: its: store: settings: allow using zephyr,settings-partition

Align with the behavior of the settings subsystem.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

rghaddab reviewed (dismissed):
Only some minor comments that you are free to take into account or not

de-nordic commented on subsys/secure_storage/Kconfig.its_store:
```diff
@@ -25,6 +46,20 @@ config SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_CUSTOM
+	  The sector size impacts the runtime behavior of ZMS and restricts the maximum
+	  ITS entry data size (which is the sector size minus ZMS and ITS overhead).
+	  Changing it will result in loss of existing data stored on a partition.
+	  It must be a multiple of the flash page size.
```
> only for a certain type of flash (that requires erase)

I would rather say "for devices that require erase". There are devices that have erase and have paging  (only for that purpose) but you can just write them without explicitly erasing first and without loosing written data in process.

Laczen reviewed (dismissed):
The proposed solution does not comply with the requirements of the psa-its standards. Why introduce the psa-its if we don't follow or adhere to the requirements ?

My change request: modify the solution to follow the psa-its standard.

tomi-font commented:
> The proposed solution does not comply with the requirements of the psa-its standards. Why introduce the psa-its if we don't follow or adhere to the requirements ?
> 
> My change request: modify the solution to follow the psa-its standard.

It has been extensively discussed and documented why we want an implementation that is not fully compliant with the official specification.
In:
- Discussions of the security working group.
- The [RFC](https://github.com/zephyrproject-rtos/zephyr/issues/75275).
- The [subsystem's documentation](https://docs.zephyrproject.org/4.0.0/services/secure_storage.html#:~:text=does%20not%20aim%20at%20full%20compliance).

This is something that relevant stakeholders have approved.

Laczen commented:
@ceolin I am directing my concerns to you as the chairperson of the security group.

By bringing in the psa-its standard into zephyr there are IMHO 2 reasons:
1. Allow standardising on the psa-its API,
2. Allow interoperability between different systems using a shared API,

The limitation of the psa-its API entries to zephyr specifications breaks reason number 2 and might result in incompatibilities between systems using a shared API. This is dangerous and might lead to not operational systems that are using a shared API. It is easy to imagine a solution where all systems would be instructed to load the key stored at 0xffffffff and the zephyr solution would fail. Cutting corners like is proposed in the solution has lead to problems in the past (e.g. the Mars Climate Orbiter incident) and should be avoided at all cost.

tomi-font commented:
> This is dangerous and might lead to not operational systems that are using a shared API

Exactly. Currently it's not safe for different users of the API that are not aware of each other. And this is is true whether the UID range is limited or not. Thus my mention of a follow-up PR that will address this issue which has always been present, since even before the introduction of the secure storage subsystem.

> It is easy to imagine a solution where all systems would be instructed to load the key stored at 0xffffffff and the zephyr solution would fail.

You have a fundamental misunderstanding here. UID values are not chosen randomly/dynamically.
It seems you need to learn about PSA Crypto, PSA Secure Storage and maybe even Mbed TLS/TF-M to get a proper understanding of the topic.

But most importantly, you need to stop trying to block legitimate contributions with blanket concerns. This is even more ridiculous when it's clear that you don't know what you're talking about.

frkv commented:
It is reasonable to ask questions and give comments in PR's, but the topics tackled here has been given ample attention both in the Zephyr Security Committee, The Zephyr Security Group in multiple meetings over a long time. The additions suggested in this PR is in my opinion well vetted and understood and accepted across the board for the people who decides what security means in Zephyr

I think we can gain much from the standardization of the API(s) in their functional form. Everything that is complicated about this in terms of security (or lack thereof) is thoroughly documented and will continue to get reasonable and rational attention in time to come

tomi-font commented on subsys/secure_storage/Kconfig.its_store:
```diff
@@ -4,12 +4,33 @@
 	depends on FLASH_HAS_DRIVER_ENABLED \
-		&& $(dt_path_enabled,$(DT_STORAGE_PARTITION)) \
-		&& $(dt_node_has_compat,$(dt_node_parent,$(DT_STORAGE_PARTITION)),fixed-partitions)
+	 && (($(dt_path_enabled,$(DT_SETTINGS_PARTITIION)) \
+	     && $(dt_node_has_compat,$(dt_node_parent,$(DT_SETTINGS_PARTITIION)),fixed-partitions))\
+	  || ($(dt_path_enabled,$(DT_STORAGE_PARTITION)) \
+	      && $(dt_node_has_compat,$(dt_node_parent,$(DT_STORAGE_PARTITION)),fixed-partitions)))
```
What would be the correct alignment for you? My logic was to make it so that the `&&` starts at the same column than where the condition it accompanies starts.

rghaddab replied:
Your logic is :
depends on A && ((B && C) || (D && E))
My concern of alignment is only on the ||
I would put it
```suggestion
		 && (($(dt_path_enabled,$(DT_SETTINGS_PARTITIION)) \
			 && $(dt_node_has_compat,$(dt_node_parent,$(DT_SETTINGS_PARTITIION)),fixed-partitions))\
			|| ($(dt_path_enabled,$(DT_STORAGE_PARTITION)) \
			    && $(dt_node_has_compat,$(dt_node_parent,$(DT_STORAGE_PARTITION)),fixed-partitions)))
```

rghaddab replied:
@tomi-font discard my comment, I tried it locally and there is no perfect way to do this without exceeding the limit of 100 column

tomi-font replied:
Yeah I forgot to say, there were actually two constraints to work under here: the 100 columns and the fact that splitting a condition (like `$(dt_node has compat,[...])` into multiple lines doesn't work (I think).

de-nordic reviewed (dismissed):
Looks OK.

ceolin commented:
> @ceolin I am directing my concerns to you as the chairperson of the security group.
> 
> By bringing in the psa-its standard into zephyr there are IMHO 2 reasons:
> 
> 1. Allow standardising on the psa-its API,
> 2. Allow interoperability between different systems using a shared API,
> 
> The limitation of the psa-its API entries to zephyr specifications breaks reason number 2 and might result in incompatibilities between systems using a shared API. This is dangerous and might lead to not operational systems that are using a shared API. It is easy to imagine a solution where all systems would be instructed to load the key stored at 0xffffffff and the zephyr solution would fail. Cutting corners like is proposed in the solution has lead to problems in the past (e.g. the Mars Climate Orbiter incident) and should be avoided at all cost.

@Laczen these are valid questions to be asked :) It is so true that we spend a reasonable amount of time debating it in the security working group. The conclusion is that the best we can do regarding clash is what is being done in this pr. As @frkv said this is going to be extensively documented / emphasized to minimize the risk.

tomi-font commented:
> @tomi-font it does not hurt to mark it as experimental for now IMHO. Lets give it some time for testing before assume it is ready for production.

For sure. In fact, [the whole subsystem is already marked as experimental](https://github.com/zephyrproject-rtos/zephyr/blob/v4.0.0/subsys/secure_storage/Kconfig#L4-L7). :slightly_smiling_face:

tomi-font commented on subsys/secure_storage/src/its/store/zms.c:
```diff
@@ -0,0 +1,123 @@
+
+static uint32_t zms_id_from(secure_storage_its_uid_t uid)
+{
+	return (uint32_t)uid.uid | (uid.caller_id << ITS_CALLER_ID_POS);
```
The result obtained from this is not used before `uid` is checked to not have any upper bits set (in `has_forbidden_bits_set()`), so here `uid.uid` can be considered to be (even less than) a `uint32_t`.

Laczen replied:
1. Please look at your code,
2. Where is it documented that this specific implementation of the ITS store module has this limitation on bits?
3. How to handle that one implementation of the ITS store module has this limitation, while another does not?
4. Why use a 64bit uid when only 30bit are supported?

tomi-font replied:
1. [subsys/secure_storage/Kconfig.its_store:24](https://github.com/zephyrproject-rtos/zephyr/pull/81235/files#diff-fb5682f6af21bb48291e3d8c7e6c7a2ba99a9601dfc2e97f4deabe459424350dR24-R25)
2. By letting the users of the API only use a well-defined range of UIDs. Will come in a follow-up PR as stated previously.
3. The 64-bit size originally comes from the specification. It could indeed be reduced when allowed values are officially defined.

tomi-font committed:
secure_storage: its: store: settings: improve debug logging

Align the debug logging with that of the ZMS-based implementation.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font commented:
Rebased to fix a conflict in `overlay-default_transform.conf` (now `overlay-transform_default.conf`).

tomi-font commented:
@Laczen Following [the contribution process](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation), I am dismissing your review as you have just nacked this PR on the premise that the implementation is not compliant with the specification, which is something that has already been discussed and approved in the security working group. Plus this subsystem is still new and experimental; improvements are to come which will make it more usable.
Adding the DNM label for one business day.
